Issue 20

aeciopires opened the pull request:
#### What this PR does / why we need it:
> **Break change 4.0.0**
* Will be used Postgresql 15.x and Zabbix 6.x.
* Tested support with PostgreSQL 15.x, TimeScaleDB 2.10.x and Zabbix 6.4.x
* Allow install zabbix-agent2 as deployment and sidecar container. More info: https://github.com/zabbix-community/helm-zabbix/issues/20
* This release changes parameter names in preparation for addressing these issues in the future and use [camelCase](https://en.wikipedia.org/wiki/Camel_case) pattern where is possible. More info: https://github.com/zabbix-community/helm-zabbix/issues/18 and https://github.com/zabbix-community/helm-zabbix/issues/21
  * ``db_access`` -> ``postgresAccess``
  * ``db_access.use_unified_secret`` -> ``postgresAccess.useUnifiedSecret``
  * ``db_access.unified_secret_name`` -> ``postgresAccess.unifiedSecretName``
  * ``db_access.unified_secret_autocreate`` -> ``postgresAccess.unifiedSecretAutoCreate``
  * ``db_access.db_server_host`` -> ``postgresAccess.host``
  * ``db_access.db_server_port`` -> ``postgresAccess.port``
  * ``db_access.postgres_user`` -> ``postgresAccess.user``
  * ``db_access.postgres_password`` -> ``postgresAccess.password``
  * ``db_access.postgres_db`` -> ``postgresAccess.database``
  * ``db_access.postgres_password_secret`` -> ``postgresAccess.passwordSecret``
  * ``db_access.postgres_password_secret_key`` -> ``postgresAccess.passwordSecretKey``
  * ``ingressroute`` -> ``ingressRoute``
  * ``postgresql.existing_claim_name`` -> ``postgresql.existingClaimName``
  * ``postgresql.storage_size`` -> ``postgresql.storageSize``
  * ``postgresql.storage_class`` -> ``postgresql.storageClass``
  * ``zabbix_image_tag`` -> ``zabbixImageTag``
  * ``zabbixagent`` -> ``zabbixAgent``
  * ``zabbixproxy`` -> ``zabbixProxy``
  * ``zabbixserver`` -> ``zabbixServer``
  * ``zabbixserver.pod_anti_affinity`` -> ``zabbixServer.podAntiAffinity``
  * ``zabbixserver.ha_nodes_autoclean`` -> ``zabbixServer.haNodesAutoClean``
  * ``zabbixserver.ha_nodes_autoclean.delete_older_than_seconds`` -> ``zabbixServer.haNodesAutoClean.deleteOlderThanSeconds``
  * ``zabbixweb`` -> ``zabbixWeb``
  * ``zabbixweb.pod_anti_affinity`` -> ``zabbixWeb.podAntiAffinity``
  * ``zabbixweb.saml_certs_secret_name`` -> ``zabbixWeb.samlCertsSecretName``
  * ``zabbixwebservice`` -> ``zabbixWebService``
  * ``zabbixwebservice.pod_anti_affinity`` -> ``zabbixWebService.podAntiAffinity``
  * ``zabbixwebservice.ignore_url_cert_errors`` -> ``zabbixWebService.ignoreURLCertErrors``

#### Which issue this PR fixes
 - This release changes parameter names in preparation for addressing these issues in the future. More info: https://github.com/zabbix-community/helm-zabbix/issues/18 and https://github.com/zabbix-community/helm-zabbix/issues/21
 - fixes #20 
 - Syncronized with PRs #33 #34 and #35

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

BGmot commented on charts/zabbix/templates/deployment-zabbix-webservice.yaml:
```diff
@@ -66,24 +66,38 @@ spec:
       tolerations:
         {{- toYaml . | nindent 8 }}
       {{- end }}
+      initContainers:
```
Won't be a conflict with line 38 if I define `extraInitContainers`?

aeciopires replied:
Hi @BGmot!

Well observed. I'll fix this.

aeciopires replied:
Hi @BGmot! 

I fixed this problem.

aeciopires commented:
Hi @BGmot and @sa-ChristianAnton!

I updated the pull request with new changes in master branch and removed the initContainer that changed the Zabbix Server address after installation.

BGmot reviewed (approved):
You put me in reviewers? Thanks -)
Ok, looks good to me. Still have to digest what you actually doing in GitHub actions -\

BGmot commented on charts/zabbix/templates/deployment-zabbix-server.yaml:
```diff
@@ -63,60 +63,55 @@ spec:
+              {{- if ne "0.0.0.0" .Values.zabbixServer.hostIP }}
+              hostIP: {{ .Values.zabbixServer.hostIP }}
               {{- end}}
             - containerPort: 10052
```
I'd remove port 10052 here as anyway nothing is listening on this port. Though it does not hurt to have it here too.

BGmot commented on charts/zabbix/templates/service.yaml:
```diff
@@ -10,27 +10,27 @@ metadata:
+      {{- if ( and (eq .Values.zabbixServer.service.type "NodePort" ) (not (empty .Values.zabbixServer.service.nodePort)) ) }}
+      nodePort: {{ .Values.zabbixServer.service.nodePort }}
       {{- end }}
     - port: 10052
```
I'd remove port 10052 here as anyway nothing is listening on this port. Though it does not hurt to have it here too.

sa-ChristianAnton commented:
Hi folks,

I have finally had some time to investigate the changes. I must admit, I am not too happy with the Zabbix Agent deployment situation yet.

Implemented in this pull request is the setup that if `.zabbixAgent.enabled` is activated, then actually several Zabbix Agents will get deployed: one as a sidecar container to the zabbix server, and one as DaemonSet on every of the nodes. Whether the one being a sidecar container is just there to "turn the Zabbix Server green in Zabbix UI" but doesn't do anything useful (actually, the metrics collected by the "Linux OS" template assigned to the "Zabbix server" Host object inside Zabbix do not deliver useful metrics, because the "Zabbix server" is no longer representing a server but just a "service". Agents running as DaemonSet are, as implemented right now, not really useful for anything as they are: there is one service loadbalancing requests between the agents, so if your cluster has more than one node, it's going to be useless metrics (measurements jumping between the nodes), and there is no good way to request metrics from these in a per-node way. The only good way to achieve this would be to use "active" method for data gathering and not to deploy service objects with the DaemonSets.

I would like to take this approach a bit further and to suggest the following:

Let's find a way to modify the Zabbix configuration (via API or via SQL) which only takes place for NEW deployments, and never modifies existing installations (as you never know what purpose the setup was). This could be achieved by extending the "init-db" job and making sure this job gets also started in single-node setups. Another idea would be to run another job at the deployment time, which checks for how long the database has been deployed and if this is below n seconds, executes its tasks, otherwise not. I don't have the perfect idea here yet.

Once we have such a job or possibility, we could do the following:

## for the "sidecar'ed Zabbix Agent":
Let's add an additional value inside the "zabbixServer" section, called maybe `zabbixServer.sidecarZabbixAgent` which is default `false`. If it is false, the above mentioned job should unassign (and clean) the "OS Linux" template from the "Zabbix server" Host. This should be the default, because it is the setup that logically makes most sense: "Zabbix server" host would then only have the "Zabbix server health" template enabled and would gather metrics for self monitoring just fine.

## for the DaemonSet Zabbix Agent
These should be configured in a way that they are able to deliver metrics to Zabbix as ACTIVE agents, by properly setting the according env variables for zabbix server host, port and agent hostname to be the node's host name. The above mentioned "init" job should then configure zabbix to activate "Agent auto registration" so that the agent instances would automatically be added to Zabbix with the proper names of the nodes.

With this feature possible to switch on or off, the user is free to use Zabbix' own helm chart which is designed to cover "Kubernetes Monitoring" and use ours just for "running Zabbix in Kubernetes", which was the original purpose. 

What do you think about this idea? Do you believe this would be feasable?

aeciopires commented:
Hi @sa-ChristianAnton !

Sorry for my absence. I was very busy in this months.

I readed all comments. 

Can we to divide the problem in some parts?

**Part 1:** in this PR, I can delete changes of zabbix-agent sidecar and the Zabbix Server. I don't have expertise with Zabbix API or database to implement your suggestions. Do you can implement this feature in another PR?

**Part 2:** in this PR I can enable/disable the zabbix-agent installation as daemonset without using external helm chart of Zabbix SIA. If disable, the user can to choice use the external helm chart of Zabbix SIA. If enable, the instalation will use our code.

**Part 3:** in this PR I can enable/disable the creation of zabbix-agent as sidecar in the Zabbix Server deployment.

What do you think? 

This way we don't lock up the changes in this PR and we have room for future improvements

Best regards

aeciopires commented:
Hi @sa-ChristianAnton!

I back of vacation.

I made the adjusts in this PR and synced with new changes in PRs #33, #34 and #35. 
Could you do the review?
Have a nive week

aeciopires commented:
Hi @sa-ChristianAnton!

Sorry to make the merge without your review and approve. 

I waited for some months and this Pull Request contains important changes for evolution of chart.
